File: api/HrManagementSystem/Shared/Consts/Permissions.cs
namespace HrManagementSystem.Shared.Consts;

public static class Permissions
{
    public static string Type { get; } = "Permissions";

    public const string ViewAddresses = "Addresses:View";
    public const string CreateAddresses = "Addresses:Create";
    public const string EditAddresses = "Addresses:Edit";
    public const string DeleteAddresses = "Addresses:Delete";

    public const string ViewAddressTypes = "AddressTypes:View";
    public const string CreateAddressTypes = "AddressTypes:Create";
    public const string EditAddressTypes = "AddressTypes:Edit";
    public const string DeleteAddressTypes = "AddressTypes:Delete";

    public const string ViewApiKeys = "ApiKeys:View";
    public const string CreateApiKeys = "ApiKeys:Create";
    public const string EditApiKeys = "ApiKeys:Edit";
    public const string DeleteApiKeys = "ApiKeys:Delete";

    public const string ViewBackups = "Backups:View";
    public const string CreateBackups = "Backups:Create";
    public const string RestoreBackups = "Backups:Restore";
    public const string DeleteBackups = "Backups:Delete";

    public const string ViewCategories = "Categories:View";
    public const string CreateCategories = "Categories:Create";
    public const string EditCategories = "Categories:Edit";
    public const string DeleteCategories = "Categories:Delete";

    public const string ViewCountries = "Countries:View";
    public const string CreateCountries = "Countries:Create";
    public const string EditCountries = "Countries:Edit";
    public const string DeleteCountries = "Countries:Delete";

    public const string ViewDistricts = "Districts:View";
    public const string CreateDistricts = "Districts:Create";
    public const string EditDistricts = "Districts:Edit";
    public const string DeleteDistricts = "Districts:Delete";

    public const string ViewChangeLogs = "ChangeLogs:View";
    public const string ViewHangfireDashboard = "Hangfire:View";
    public const string ManageDatabaseViews = "DatabaseViews:Manage";

    public const string ViewLocalizations = "Localizations:View";
    public const string CreateLocalizations = "Localizations:Create";
    public const string EditLocalizations = "Localizations:Edit";
    public const string DeleteLocalizations = "Localizations:Delete";

    public const string ViewReportsCategories = "ReportsCategories:View";
    public const string CreateReportsCategories = "ReportsCategories:Create";
    public const string EditReportsCategories = "ReportsCategories:Edit";
    public const string DeleteReportsCategories = "ReportsCategories:Delete";

    public const string ViewRoles = "Roles:View";
    public const string CreateRoles = "Roles:Create";
    public const string EditRoles = "Roles:Edit";
    public const string DeleteRoles = "Roles:Delete";

    public const string ViewStates = "States:View";
    public const string CreateStates = "States:Create";
    public const string EditStates = "States:Edit";
    public const string DeleteStates = "States:Delete";

    public const string ViewSubCategories = "SubCategories:View";
    public const string CreateSubCategories = "SubCategories:Create";
    public const string EditSubCategories = "SubCategories:Edit";
    public const string DeleteSubCategories = "SubCategories:Delete";

    // Kanban Boards Permissions
    public const string ViewKanbanBoards = "KanbanBoards:View";
    public const string CreateKanbanBoards = "KanbanBoards:Create";
    public const string EditKanbanBoards = "KanbanBoards:Edit";
    public const string DeleteKanbanBoards = "KanbanBoards:Delete";

    // Kanban Columns Permissions
    public const string ViewKanbanColumns = "KanbanColumns:View";
    public const string CreateKanbanColumns = "KanbanColumns:Create";
    public const string EditKanbanColumns = "KanbanColumns:Edit";
    public const string DeleteKanbanColumns = "KanbanColumns:Delete";

    // Kanban Cards Permissions
    public const string ViewKanbanCards = "KanbanCards:View";
    public const string CreateKanbanCards = "KanbanCards:Create";
    public const string EditKanbanCards = "KanbanCards:Edit";
    public const string DeleteKanbanCards = "KanbanCards:Delete";

    // Kanban Card Assignees Permissions
    public const string ViewKanbanCardAssignees = "KanbanCardAssignees:View";
    public const string CreateKanbanCardAssignees = "KanbanCardAssignees:Create";
    public const string EditKanbanCardAssignees = "KanbanCardAssignees:Edit";
    public const string DeleteKanbanCardAssignees = "KanbanCardAssignees:Delete";

    // Kanban Card Attachments Permissions
    public const string ViewKanbanCardAttachments = "KanbanCardAttachments:View";
    public const string CreateKanbanCardAttachments = "KanbanCardAttachments:Create";
    public const string EditKanbanCardAttachments = "KanbanCardAttachments:Edit";
    public const string DeleteKanbanCardAttachments = "KanbanCardAttachments:Delete";

    // Board Task Attachments Permissions
    public const string ViewBoardTaskAttachments = "BoardTaskAttachments:View";
    public const string CreateBoardTaskAttachments = "BoardTaskAttachments:Create";
    public const string EditBoardTaskAttachments = "BoardTaskAttachments:Edit";
    public const string DeleteBoardTaskAttachments = "BoardTaskAttachments:Delete";

    // Kanban Card Comments Permissions
    public const string ViewKanbanCardComments = "KanbanCardComments:View";
    public const string CreateKanbanCardComments = "KanbanCardComments:Create";
    public const string EditKanbanCardComments = "KanbanCardComments:Edit";
    public const string DeleteKanbanCardComments = "KanbanCardComments:Delete";

    // Board Task Comments Permissions
    public const string ViewBoardTaskComments = "BoardTaskComments:View";
    public const string CreateBoardTaskComments = "BoardTaskComments:Create";
    public const string EditBoardTaskComments = "BoardTaskComments:Edit";
    public const string DeleteBoardTaskComments = "BoardTaskComments:Delete";

    // Kanban Labels Permissions
    public const string ViewKanbanLabels = "KanbanLabels:View";
    public const string CreateKanbanLabels = "KanbanLabels:Create";
    public const string EditKanbanLabels = "KanbanLabels:Edit";
    public const string DeleteKanbanLabels = "KanbanLabels:Delete";

    // Kanban Board Members Permissions
    public const string ViewKanbanBoardMembers = "KanbanBoardMembers:View";
    public const string CreateKanbanBoardMembers = "KanbanBoardMembers:Create";
    public const string EditKanbanBoardMembers = "KanbanBoardMembers:Edit";
    public const string DeleteKanbanBoardMembers = "KanbanBoardMembers:Delete";

    public const string ViewUsers = "Users:View";
    public const string CreateUsers = "Users:Create";
    public const string EditUsers = "Users:Edit";
    public const string DeleteUsers = "Users:Delete";

    // Chat User Permissions
    public const string ViewChatUsers = "ChatUsers:View";
    public const string CreateChatUsers = "ChatUsers:Create";
    public const string EditChatUsers = "ChatUsers:Edit";
    public const string DeleteChatUsers = "ChatUsers:Delete";

    // Conversation Permissions
    public const string ViewConversations = "Conversations:View";
    public const string CreateConversations = "Conversations:Create";
    public const string EditConversations = "Conversations:Edit";
    public const string DeleteConversations = "Conversations:Delete";

    // Message Permissions
    public const string ViewMessages = "Messages:View";
    public const string CreateMessages = "Messages:Create";
    public const string EditMessages = "Messages:Edit";
    public const string DeleteMessages = "Messages:Delete";

    // Chat General Permissions
    public const string AccessChat = "Chat:Access";
    public const string ModerateChat = "Chat:Moderate";
    public const string ViewChatAnalytics = "Chat:ViewAnalytics";

    public static IReadOnlyList<string> GetAllPermissions() =>
        typeof(Permissions).GetFields()
            .Select(field => field.GetValue(null))
            .OfType<string>()
            .ToList();

}

File: web-next/src/lib/auth/permissions.ts
// Mirrors api/HrManagementSystem/Shared/Consts/Permissions.cs; parity is enforced by a test.
export const permissions = {
  ViewAddresses: "Addresses:View",
  CreateAddresses: "Addresses:Create",
  EditAddresses: "Addresses:Edit",
  DeleteAddresses: "Addresses:Delete",
  ViewAddressTypes: "AddressTypes:View",
  CreateAddressTypes: "AddressTypes:Create",
  EditAddressTypes: "AddressTypes:Edit",
  DeleteAddressTypes: "AddressTypes:Delete",
  ViewApiKeys: "ApiKeys:View",
  CreateApiKeys: "ApiKeys:Create",
  EditApiKeys: "ApiKeys:Edit",
  DeleteApiKeys: "ApiKeys:Delete",
  ViewBackups: "Backups:View",
  CreateBackups: "Backups:Create",
  RestoreBackups: "Backups:Restore",
  DeleteBackups: "Backups:Delete",
  ViewCategories: "Categories:View",
  CreateCategories: "Categories:Create",
  EditCategories: "Categories:Edit",
  DeleteCategories: "Categories:Delete",
  ViewCountries: "Countries:View",
  CreateCountries: "Countries:Create",
  EditCountries: "Countries:Edit",
  DeleteCountries: "Countries:Delete",
  ViewDistricts: "Districts:View",
  CreateDistricts: "Districts:Create",
  EditDistricts: "Districts:Edit",
  DeleteDistricts: "Districts:Delete",
  ViewChangeLogs: "ChangeLogs:View",
  ViewHangfireDashboard: "Hangfire:View",
  ManageDatabaseViews: "DatabaseViews:Manage",
  ViewLocalizations: "Localizations:View",
  CreateLocalizations: "Localizations:Create",
  EditLocalizations: "Localizations:Edit",
  DeleteLocalizations: "Localizations:Delete",
  ViewReportsCategories: "ReportsCategories:View",
  CreateReportsCategories: "ReportsCategories:Create",
  EditReportsCategories: "ReportsCategories:Edit",
  DeleteReportsCategories: "ReportsCategories:Delete",
  ViewRoles: "Roles:View",
  CreateRoles: "Roles:Create",
  EditRoles: "Roles:Edit",
  DeleteRoles: "Roles:Delete",
  ViewStates: "States:View",
  CreateStates: "States:Create",
  EditStates: "States:Edit",
  DeleteStates: "States:Delete",
  ViewSubCategories: "SubCategories:View",
  CreateSubCategories: "SubCategories:Create",
  EditSubCategories: "SubCategories:Edit",
  DeleteSubCategories: "SubCategories:Delete",
  ViewKanbanBoards: "KanbanBoards:View",
  CreateKanbanBoards: "KanbanBoards:Create",
  EditKanbanBoards: "KanbanBoards:Edit",
  DeleteKanbanBoards: "KanbanBoards:Delete",
  ViewKanbanColumns: "KanbanColumns:View",
  CreateKanbanColumns: "KanbanColumns:Create",
  EditKanbanColumns: "KanbanColumns:Edit",
  DeleteKanbanColumns: "KanbanColumns:Delete",
  ViewKanbanCards: "KanbanCards:View",
  CreateKanbanCards: "KanbanCards:Create",
  EditKanbanCards: "KanbanCards:Edit",
  DeleteKanbanCards: "KanbanCards:Delete",
  ViewKanbanCardAssignees: "KanbanCardAssignees:View",
  CreateKanbanCardAssignees: "KanbanCardAssignees:Create",
  EditKanbanCardAssignees: "KanbanCardAssignees:Edit",
  DeleteKanbanCardAssignees: "KanbanCardAssignees:Delete",
  ViewKanbanCardAttachments: "KanbanCardAttachments:View",
  CreateKanbanCardAttachments: "KanbanCardAttachments:Create",
  EditKanbanCardAttachments: "KanbanCardAttachments:Edit",
  DeleteKanbanCardAttachments: "KanbanCardAttachments:Delete",
  ViewBoardTaskAttachments: "BoardTaskAttachments:View",
  CreateBoardTaskAttachments: "BoardTaskAttachments:Create",
  EditBoardTaskAttachments: "BoardTaskAttachments:Edit",
  DeleteBoardTaskAttachments: "BoardTaskAttachments:Delete",
  ViewKanbanCardComments: "KanbanCardComments:View",
  CreateKanbanCardComments: "KanbanCardComments:Create",
  EditKanbanCardComments: "KanbanCardComments:Edit",
  DeleteKanbanCardComments: "KanbanCardComments:Delete",
  ViewBoardTaskComments: "BoardTaskComments:View",
  CreateBoardTaskComments: "BoardTaskComments:Create",
  EditBoardTaskComments: "BoardTaskComments:Edit",
  DeleteBoardTaskComments: "BoardTaskComments:Delete",
  ViewKanbanLabels: "KanbanLabels:View",
  CreateKanbanLabels: "KanbanLabels:Create",
  EditKanbanLabels: "KanbanLabels:Edit",
  DeleteKanbanLabels: "KanbanLabels:Delete",
  ViewKanbanBoardMembers: "KanbanBoardMembers:View",
  CreateKanbanBoardMembers: "KanbanBoardMembers:Create",
  EditKanbanBoardMembers: "KanbanBoardMembers:Edit",
  DeleteKanbanBoardMembers: "KanbanBoardMembers:Delete",
  ViewUsers: "Users:View",
  CreateUsers: "Users:Create",
  EditUsers: "Users:Edit",
  DeleteUsers: "Users:Delete",
  ViewChatUsers: "ChatUsers:View",
  CreateChatUsers: "ChatUsers:Create",
  EditChatUsers: "ChatUsers:Edit",
  DeleteChatUsers: "ChatUsers:Delete",
  ViewConversations: "Conversations:View",
  CreateConversations: "Conversations:Create",
  EditConversations: "Conversations:Edit",
  DeleteConversations: "Conversations:Delete",
  ViewMessages: "Messages:View",
  CreateMessages: "Messages:Create",
  EditMessages: "Messages:Edit",
  DeleteMessages: "Messages:Delete",
  AccessChat: "Chat:Access",
  ModerateChat: "Chat:Moderate",
  ViewChatAnalytics: "Chat:ViewAnalytics",
} as const;

export type PermissionString =
  (typeof permissions)[keyof typeof permissions];

// Manually defined for proper type inference (TypeScript can't infer from template literals in unions)
export type PermissionModule =
  | "Addresses"
  | "AddressTypes"
  | "ApiKeys"
  | "Backups"
  | "Categories"
  | "Countries"
  | "Districts"
  | "ChangeLogs"
  | "Hangfire"
  | "DatabaseViews"
  | "Localizations"
  | "ReportsCategories"
  | "Roles"
  | "States"
  | "SubCategories"
  | "KanbanBoards"
  | "KanbanColumns"
  | "KanbanCards"
  | "KanbanCardAssignees"
  | "KanbanCardAttachments"
  | "BoardTaskAttachments"
  | "KanbanCardComments"
  | "BoardTaskComments"
  | "KanbanLabels"
  | "KanbanBoardMembers"
  | "Users"
  | "ChatUsers"
  | "Conversations"
  | "Messages"
  | "Chat";

export const getAllPermissions = (): PermissionString[] =>
  Object.values(permissions);

export const getPermissionModule = (
  permission: PermissionString
): PermissionModule => {
  const parts = permission.split(":");
  if (parts.length !== 2 || !parts[0] || !parts[1]) {
    throw new Error(`Invalid permission format: "${permission}". Expected "Module:Action".`);
  }
  return parts[0] as PermissionModule;
};

export const getAllPermissionModules = (): PermissionModule[] => [
  ...new Set(getAllPermissions().map(getPermissionModule)),
];

// Cache permission sets for performance (avoids repeated array iterations)
const permissionSetCache = new WeakMap<readonly string[], Set<string>>();

function getPermissionSet(userPermissions: readonly string[]): Set<string> {
  let cached = permissionSetCache.get(userPermissions);
  if (!cached) {
    cached = new Set(userPermissions);
    permissionSetCache.set(userPermissions, cached);
  }
  return cached;
}

export const hasPermission = (
  userPermissions: readonly string[],
  permission: PermissionString
): boolean => getPermissionSet(userPermissions).has(permission);

export const hasAnyPermission = (
  userPermissions: readonly string[],
  requiredPermissions: readonly PermissionString[]
): boolean => {
  const permissionSet = getPermissionSet(userPermissions);
  return requiredPermissions.some((permission) => permissionSet.has(permission));
};

export const hasAllPermissions = (
  userPermissions: readonly string[],
  requiredPermissions: readonly PermissionString[]
): boolean => {
  const permissionSet = getPermissionSet(userPermissions);
  return requiredPermissions.every((permission) => permissionSet.has(permission));
};
